gap_linux.go

```diff
@@ -284,25 +291,73 @@ func (a *Adapter) Connect(address Addresser, params ConnectionParams) (*Device,
 		return nil, err
 	}
 
+	device := &Device{
+		device:  dev,
+		adapter: a,
+		address: &adr,
+	}
+	device.ctx, device.cancel = context.WithCancel(context.Background())
+	device.watchForConnect() // Set this up before we trigger a conneciton so we can capture the connect event
+
 	if !dev.Properties.Connected {
 		// Not yet connected, so do it now.
 		// The properties have just been read so this is fresh data.
 		err := dev.Connect()
 		if err != nil {
+			device.cancel() // cancel our watcher routine
 			return nil, err
 		}
 	}
 
-	// TODO: a proper async callback.
-	a.connectHandler(nil, true)
-
-	return &Device{
-		device: dev,
-	}, nil
+	return device, nil
 }
 
 // Disconnect from the BLE device. This method is non-blocking and does not
 // wait until the connection is fully gone.
 func (d *Device) Disconnect() error {
+	// we don't call our cancel function here, instead we wait for the property change in `watchForConnect` and cancel things then
 	return d.device.Disconnect()
 }
+
+// watchForConnect watches for a signal from the bluez device interface that indicates a Connection/Disconnection.
+//
+// We can add extra signals to watch for here,
+// see https://git.kernel.org/pub/scm/bluetooth/bluez.git/tree/doc/device-api.txt, for a full list
+func (d *Device) watchForConnect() error {
+	var err error
+	d.propchanged, err = d.device.WatchProperties()
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		for {
+			select {
+			case changed := <-d.propchanged:
+
+				if changed == nil {
+					d.cancel()
+					return
+				}
+
+				switch changed.Name {
+				case "Connected":
+					// Send off a notification indicating we have connected or disconnected
+					d.adapter.connectHandler(d.address, d.device.Properties.Connected)
```

Comment:
The connected state also comes through in the `propchanged` event under `changed.Value`. However, it seemed more appropriate to use the actual property it is notifying us of changes about. Either way works fine in my testing.